src/ir/term/mod.rs

```diff
@@ -1252,175 +1258,225 @@ impl Value {
+        }
+        Op::Select => {
+            let a = vs.get(&c.cs[0]).unwrap().as_array().clone();
+            let i = vs.get(&c.cs[1]).unwrap();
+            a.clone().select(i)
+        },
+        Op::Map(op) => {
+            let arg_cnt = c.cs.len();
+            let arr_size = vs.get(&c.cs[0]).unwrap().as_array().size;
+            let mut term_vecs = vec![Vec::new(); arr_size];
+            //2D vector: term_vecs[i] will store a vector of all the i-th index 
+            //  entries of the array arguments
+
+            //Value::BitVector(BitVector::new(uint.into(),width,))
+            for i in 0..arg_cnt {
+                let arr = vs.get(&c.cs[i]).unwrap().as_array().clone();
+                for j in 0..arr_size {
+                    let jval = &Value::BitVector(BitVector::new(Integer::from(j), 32));
+                    let term = leaf_term(Op::Const(arr.clone().select(jval)));
+                    term_vecs[j].push(term);
                 }
-            }),
-            Op::PfNaryOp(o) => Value::Field({
-                let mut xs = c.cs.iter().map(|c| vs.get(c).unwrap().as_pf().clone());
-                let f = xs.next().unwrap();
-                xs.fold(
-                    f,
-                    match o {
-                        PfNaryOp::Add => std::ops::Add::add,
-                        PfNaryOp::Mul => std::ops::Mul::mul,
-                    },
-                )
-            }),
-            Op::UbvToPf(m) => Value::Field({
-                let a = vs.get(&c.cs[0]).unwrap().as_bv().clone();
-                field::FieldElem::new(a.uint().clone(), m.clone())
-            }),
-            // tuple
-            Op::Tuple => Value::Tuple(c.cs.iter().map(|c| vs.get(c).unwrap().clone()).collect()),
-            Op::Field(i) => {
-                let t = vs.get(&c.cs[0]).unwrap().as_tuple();
-                assert!(i < &t.len(), "{} out of bounds for {}", i, c.cs[0]);
-                t[*i].clone()
-            }
-            Op::Update(i) => {
-                let mut t = Vec::from(vs.get(&c.cs[0]).unwrap().as_tuple()).into_boxed_slice();
-                assert!(i < &t.len(), "{} out of bounds for {}", i, c.cs[0]);
-                let e = vs.get(&c.cs[1]).unwrap().clone();
-                assert_eq!(t[*i].sort(), e.sort());
-                t[*i] = e;
-                Value::Tuple(t)
-            }
-            // array
-            Op::Store => {
-                let a = vs.get(&c.cs[0]).unwrap().as_array().clone();
-                let i = vs.get(&c.cs[1]).unwrap().clone();
-                let v = vs.get(&c.cs[2]).unwrap().clone();
-                Value::Array(a.clone().store(i, v))
             }
-            Op::Select => {
-                let a = vs.get(&c.cs[0]).unwrap().as_array().clone();
-                let i = vs.get(&c.cs[1]).unwrap();
-                a.clone().select(i)
+            //Cloning the first arg just for formatting
+            let mut res = vs.get(&c.cs[0]).unwrap().as_array().clone();
```

Comment:
We could create a new array term by using: `Array::default()` from src/ir/term/mod.rs. We would just need to get the key sort, value sort, and size from vs. 

@alex-ozdemir thoughts on this? Since we are already replacing all of the values in the array, either approach is logically the same.

Reply:
Either approach is good in my mind.

Reply:
I've though about this more and it looks like I should use Array::default() instead of cloning. If I use the cloning approach, it causes the output array term to have an incorrect value type in some cases.

Reply:
I'm trying to use the check() function from src/ir/term/ty.rs to figure out the type to use for Array::default(), but I am running into an issue where check(&c) is returning Sort::bool when "c" is a Map term.
I looked a bit further into this bug and it looks like the type of the Map term is stored as a Boolean in TREM_TYPES, so the check_raw() is returning in its first 3 lines:
```
if let Some(s) = TERM_TYPES.read().unwrap().get(&t.to_weak()) {
        return Ok(s.clone());
    }
```
Is there a way I can fix this issue?

Reply:
`TERM_TYPES` is just a cache for `check` (and friends), so the relevant question is when we put the wrong sort into the map. I'd just print out every store to the map, and search for the bad insertion.

Reply:
Looks like there was a bug in my check_rec() implementation. I've made the fix and the new approach for initializing the array is working as intended.